router/internal/expr/expr.go

```diff
@@ -0,0 +1,169 @@
+func (r RequestHeaders) Get(key string) string {
+	return http.Header(r).Get(key)
+}
+
+// LoadRequest loads the request object into the context.
+func (r *RequestRootContext) LoadRequest(req *http.Request) {
+	if req == nil {
+		return
+	}
+
+	r.Request.Header = RequestHeaders(req.Header)
+
+	m, _ := url.ParseQuery(req.URL.RawQuery)
+	qv := make(map[string]string, len(m))
+
+	for k := range m {
+		qv[k] = m.Get(k)
+	}
+
+	r.Request.URL = RequestURL{
+		Method: req.Method,
+		Scheme: req.URL.Scheme,
+		Host:   req.URL.Host,
+		Path:   req.URL.Path,
+		Query:  qv,
+	}
+}
+
+type RequestAuth struct {
+	IsAuthenticated bool           `expr:"isAuthenticated"`
+	Type            string         `expr:"type"`
+	Claims          map[string]any `expr:"claims"`
+	Scopes          []string       `expr:"scopes"`
+}
+
+// LoadAuth loads the authentication context into the request object.
+// Must only be called when the authentication was successful.
+func (r *RequestRootContext) LoadAuth(ctx context.Context) {
+	authCtx := authentication.FromContext(ctx)
+	if authCtx == nil {
+		return
+	}
+
+	r.Request.Auth.Type = authCtx.Authenticator()
+	r.Request.Auth.IsAuthenticated = true
+	r.Request.Auth.Claims = authCtx.Claims()
+	r.Request.Auth.Scopes = authCtx.Scopes()
+}
+
+// CompileBoolExpression compiles an expression and returns the program. It is used for expressions that return bool.
+// The exprContext is used to provide the context for the expression evaluation.
+func CompileBoolExpression(s string) (*vm.Program, error) {
+	v, err := expr.Compile(s, expr.Env(RequestRootContext{}), expr.AsBool())
+	if err != nil {
+		return nil, handleExpressionError(err)
+	}
+
+	return v, nil
+}
+
+// CompileStringExpression compiles an expression and returns the program. It is used for expressions that return strings
+// The exprContext is used to provide the context for the expression evaluation.
+func CompileStringExpression(s string) (*vm.Program, error) {
+	v, err := expr.Compile(s, expr.Env(RequestRootContext{}), expr.AsKind(reflect.String))
+	if err != nil {
+		return nil, handleExpressionError(err)
+	}
+
+	return v, nil
+}
+
+// ResolveStringExpression evaluates the expression and returns the result as a string. The exprContext is used to
+// provide the context for the expression evaluation.
+func ResolveStringExpression(vm *vm.Program, ctx RequestRootContext) (string, error) {
+	r, err := expr.Run(vm, ctx)
+	if err != nil {
+		return "", handleExpressionError(err)
+	}
+
+	switch v := r.(type) {
+	case string:
+		return v, nil
+	default:
+		return "", fmt.Errorf("expected string, got %T", r)
+	}
+}
+
+// ResolveBoolExpression evaluates the expression and returns the result as a bool. The exprContext is used to
+// provide the context for the expression evaluation.
+func ResolveBoolExpression(vm *vm.Program, ctx RequestRootContext) (bool, error) {
+	if vm == nil {
+		return false, nil
+	}
+
+	r, err := expr.Run(vm, ctx)
```

Comment:
It appears to me as if expr.Run is not thread safe.
I think we need to have a mutex around it, and potentially compile one expr per CPU and then have a slice of expressions (one per CPU) where we can pick one vm from a semaphore.

Reply:
Why do you think it needs to be thread safe? You suggestions sounds like a big pre optimization that isn't required at this point.

Reply:
I think it needs to be thread safe because you have one single operation blocker which gets called by many requests in different goroutines.

Reply:
It is thread safe https://github.com/expr-lang/expr/issues/683#issuecomment-2213958821